lib/Dialect/Dialect.cpp

```diff
@@ -19,17 +19,119 @@
+
+  std::atomic<LLVMContext *> m_llvmContext = nullptr;
+  DialectContext *m_dialectContext = nullptr;
+  CurrentContextCache **m_pprev;
+  CurrentContextCache *m_next = nullptr;
+
+  CurrentContextCache() {
+    auto &map = ContextMap::get();
+    auto lock = std::lock_guard(map.m_mutex);
+    m_next = map.m_caches;
+    if (m_next)
+      m_next->m_pprev = &m_next;
+    m_pprev = &map.m_caches;
+    map.m_caches = this;
+  }
+
+  ~CurrentContextCache() {
+    auto &map = ContextMap::get();
+    auto lock = std::lock_guard(map.m_mutex);
+    if (m_next)
+      m_next->m_pprev = m_pprev;
+    *m_pprev = m_next;
+  }
+
+public:
+  static DialectContext *get(LLVMContext *llvmContext) {
+    assert(llvmContext != nullptr);
+    static thread_local CurrentContextCache cache;
+    if (cache.m_llvmContext.load(std::memory_order_relaxed) != llvmContext) {
+      auto &map = ContextMap::get();
+      auto lock = std::lock_guard(map.m_mutex);
+      cache.m_llvmContext.store(llvmContext, std::memory_order_relaxed);
+      cache.m_dialectContext = map.m_map.lookup(llvmContext);
+    }
+    return cache.m_dialectContext;
+  }
+};
+
+void ContextMap::insert(LLVMContext *llvmContext,
+                        DialectContext *dialectContext) {
+  auto lock = std::lock_guard(m_mutex);
+  bool inserted = m_map.try_emplace(llvmContext, dialectContext).second;
+  (void)inserted;
+  assert(inserted);
+}
+
+void ContextMap::remove(LLVMContext *llvmContext,
+                        DialectContext *dialectContext) {
+  auto lock = std::lock_guard(m_mutex);
+  assert(m_map.lookup(llvmContext) == dialectContext);
+  m_map.erase(llvmContext);
+
+  // Remove any stale per-thread cache entries.
+  //
+  // This is called when llvmContext still exists, and our thread destroys it
+  // (or at least detaches the dialectContext). No other thread can legitimately
+  // attempt to do anything with the same llvmContext at the same time.
+  //
+  // However, another thread may have previously used llvmContext and still
+  // see it in its cache. We need to null out those cache entries in case a
+  // new LLVMContext is created at the exact same address.
+  //
+  // The other thread may race us in an attempt to start using a different
+  // context. All *writes* to CurrentContextCache::m_llvmContext are guarded by
+  // ContextMap::m_mutex. But there is still a race between
+  //
+  //  1. Our thread writing to m_llvmContext and
+  //  2. The other thread checking m_llvmContext from CurrentContextCache::get
+  //
+  // This race is formally undefined behavior, so we use std::atomic. Relaxed
+  // memory orderings are sufficient because the other thread will not write to
+  // m_llvmContext without locking the m_mutex first.
+  for (CurrentContextCache *cache = m_caches; cache; cache = cache->m_next) {
+    if (cache->m_llvmContext.load(std::memory_order_relaxed) == llvmContext)
+      cache->m_llvmContext.store(nullptr, std::memory_order_relaxed);
+  }
+}
```

Comment:
The `m_mutex` is locked all the way through here already, so I'm not sure what another mutex would bring?

> How are we ensuring that another thread sees our store in his load soon enough?

There is only one other thread that can load from `cache->m_llvmContext`: the one that owns `cache`.

If that thread loads `cache->m_llvmContext` before we release the `m_mutex` lock, then it either sees `nullptr` or it sees `llvmContext`. The other thread can't be looking for either of these values (we can never look for `nullptr`, and it would be incorrect for the other thread to try to use an LLVMContext that's currently in the process of being destroyed), so the path taken by the other thread doesn't depend on whether it sees our store or not. Either way, it goes down the path where it locks the `m_mutex`. That's all the synchronization we need.

Reply:
I missed the mutex at the very top :)

Still -- as the other thread does not care about the mutex (as long as it sees a cache match), where does the guarantee come from that our store of the nullptr value is seen by another thread after the context has been destroyed and recreated?

So:

```
T1 (we):      lock -> store nullptr -> release lock -> destroy context
T2 (them):                                                               re-create context -> load  
```

Reply:
> Note, this all leans on the fact that the usual rules for LLVMContexts still apply: You can't use them simultaneously from multiple threads (at least not in a read/write fashion, and destroying them counts as a write).

So is this essentially the "relying on external synchronization" from above -- there needs to be some sort of synchronization between the destruction of the context as seen from our thread, and the creation as observed from the other thread?

Reply:
If T2 re-creates a DialectContext for the same LLVMContext, there has to have been some application-level synchronization to make that correct.

If T2 is able to re-create a context at the same address, there has to have been some synchronization via the malloc implementation.

There may be some dragons in the questions of what level of synchronization the standard actually guarantees via malloc/free/new/delete. But clearly, allocating an LLVMContext at the same address needs *some* level of happens-after behavior, and then `get()` is only allowed to be called after the DialectContext has been created for the new LLVMContext, and the DialectContext constructor locks the `m_mutex`, and **that** certainly guarantees that the new value of `m_llvmContext` has become visible to the other thread.

I've extended the comments a bit further.

Reply:
Thanks, makes sense!

Reply:
Note that the new context can also live on the same stack address, in which case there has been no malloc synchronization. But still there needs to be application-level synchronization ensuring that accessing the context from T1 here and from T2 later is ok.